### DISCUSSION TITLE
Customize Calendar

### DISCUSSION THREAD

**Discussion by arunnbabu81** [buggyj/tw5-tools#13] [General]
It's not related to tiddlyclip. Please move it to the relevant page.
I was using your Calendar plug-in one of my wiki. 
I need to align the calendar in both month and year view to the center.

<img width="1779" alt="image" src="https://user-images.githubusercontent.com/67494083/181361935-a6bf550c-ca0f-4a63-8d1c-5ae93181c230.png">

<img width="1769" alt="image" src="https://user-images.githubusercontent.com/67494083/181361985-ad4a9673-c0d8-4b7e-a3d2-c13b1634917b.png">

<img width="1840" alt="image" src="https://user-images.githubusercontent.com/67494083/181363186-94121edd-ead3-4f09-96e8-201da04bbb68.png">

<img width="1802" alt="image" src="https://user-images.githubusercontent.com/67494083/181363216-c4c0ccd3-2205-49af-ac89-5df9a91bb74e.png">

```

\whitespace trim
<div style="overflow-x:auto;width:70%;text-align:center;margin-left: auto;
  margin-right: auto;" class="tc-tagged-byam">
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="nomatch">
  <$button set="$:/temp/plugins/bj/Calendar" setTo="Year">Year</$button>
  <<calendar "" "">>
</$reveal>
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="match">
  <$button set="$:/temp/plugins/bj/Calendar" setTo="Month">Month</$button>
<<calendar 2014 "">>
</$reveal>
</div>
```
```

\rules only filteredtranscludeinline transcludeinline macrodef macrocallinline

.tc-tagged-byam {
max-width: 100%;
font-size: 1.0em;
background: <<colour code-background>>;
padding: 1em;
margin: 0.5em;
border-radius: 8px;
border: 1px solid <<colour tag-background >>;
}

```

**Comment by buggyj** [comment]
try this:

```
@@display: flex;justify-content: center;
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="nomatch">
  <$button set="$:/temp/plugins/bj/Calendar" setTo="Year">Year</$button>
  <<calendar "" "">>
</$reveal>
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="match">
  <$button set="$:/temp/plugins/bj/Calendar" setTo="Month">Month</$button>
<<calendar 2014 "">>
</$reveal>
@@
```

**Reply by arunnbabu81** [reply]
It works. Thank you.

Can you take a look at this issue also - https://github.com/buggyj/tw5-tools/issues/9

**Comment by arunnbabu81** [comment]
@buggyj 

I was using the Diary macro. I want to add a field called journal-date to all Journal tiddlers being created from the Calendar with its contents as shown below `journal-date=<<now "YYYY0MM0DD0hh0mm0ss000">>`  How to add this ?

Also can the `calendar macro` also show links for each days Journal tiddlers like in the `diary macro`. This is because I can create a Journal tiddler by just clicking on the links for each day in the Calendar made eusing `diary macro`. But the Calendar created using `calendar macro` only shows the already created Journal tiddlers as links - we can't create new Journal tiddlers from the Calendar created using `calendar macro`

**Reply by buggyj** [reply]
you could do this:
```

\define calendarplus(y)
<$ifnew fields="""{"tags":"Journal ","journal-date":"$y$"}"""><$macrorefresh $name="calendarbase" year="2022" month="" opts="diary" $refresh="calendarrefresh"/></$ifnew>
\end

<$macrocall $name=calendarplus y=<<now "YYYY0MM0DD0hh0mm0ss000">>/>
```
but the date/time will be when the macro is rendered not when you click.

With regards to diary/calendar creating new journal tiddlers: I do not understand why you cannot use the diary macro??

**Reply by arunnbabu81** [reply]
> but the date/time will be when the macro is rendered not when you click.

1. Is there any way to get the `journal-date` field value based on the date (title) of the Journal tiddler created, not the date in which it was created.

   Eric schulman's Calendar creates Journal tiddlers with `journal-date` field value corresponding to the date (title) of Journal tiddler, not the date in which the Journal tiddler is created.

   https://tiddlytools.com/timer.html#TiddlyTools%2FTime%2FJournalListChanges
   https://tiddlytools.com/timer.html#TiddlyTools%2FTime%2FCalendar

   But his Calendar is slow to load at the first time, thats why I was trying out your Calendar.

2. How to toggle between the year and month views  using the new `calendarplus` macro 
    I tried this code. Its not toggling between month and year view.
```
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="nomatch">
  <$button set="$:/temp/plugins/bj/Calendar" setTo="Year">Year</$button>
  <$macrocall $name=calendarplus y=<<now "YYYY0MM0DD0hh0mm0ss000">>/>
</$reveal>
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="match">
  <$button set="$:/temp/plugins/bj/Calendar" setTo="Month">Month</$button>
<$macrocall $name=calendarplus y=<<now "YYYY0MM0DD0hh0mm0ss000">>/>
</$reveal>
```

> With regards to diary/calendar creating new journal tiddlers: I do not understand why you cannot use the diary macro??

Ya, that works. I missed that point.

**Comment by buggyj** [comment]
Alternativity do this
```

\define actions()
<$action-sendmessage $message="tm-new-tiddler" title=<<event-paramObject-title>> time=<<now>>/>
\end
<$messagecatcher $tm-new-tiddler=<<actions>>>
<<diary>>
</$messagecatcher>

you will need this modified file:

`
[ifnew.js.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/9228978/ifnew.js.json.zip)

**Reply by arunnbabu81** [reply]
> 

The zip file is not downloadable

**Reply by buggyj** [reply]
[ifnew.js.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/9229385/ifnew.js.json.zip)

**Reply by arunnbabu81** [reply]
I have added a demo here - https://multicolumn-krystal.tiddlyhost.com/

<img width="1892" alt="image" src="https://user-images.githubusercontent.com/67494083/182042586-f18d1ad5-9fd4-426b-8969-b3d8f76def19.png">
Click on the `Journals` link in the topbar to see the demo.
I have listed 3 methods used to create journal tiddlers.

1. First column uses, the `calendar plus` macro shared by you. It create Journal tiddlers correctly except that the `journal-date` field shows the created date. For example, if I create a Journal tiddler for August 19th today, the `journal-date` field is showing todays date, not the date of August 19. Thats not what I want.
2. Second column uses the recently shared code. It doesnt work well. The tittle and journal-date field are not as I want it to be.
  
> Alternativity do this
> 
> ```
> 
> \define actions()
> <$action-sendmessage $message="tm-new-tiddler" title=<<event-paramObject-title>> time=<<now>>/>
> \end
> <$messagecatcher $tm-new-tiddler=<<actions>>>
> <<diary>>
> </$messagecatcher>
> 
> you will need this modified file:
> 
> `
> [ifnew.js.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/9228978/ifnew.js.json.zip)
> ```


  3. Third column shows Eric Calendar, which creates Journal tiddlers with`journal-date` field showing date same as that of the title of the Journal tiddler. For example, if I created a Journal tiddler for August 18th today, the journal-date field shows date of August 18 not todays date. Thats what I want. But this Calendar takes time to load when its opened for the first time

**Reply by buggyj** [reply]
you have not included the new version of ifnew.js in your example wiki - 

[ifnew.js.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/9229654/ifnew.js.json.zip)

**Reply by arunnbabu81** [reply]
> 

I had drag and dropped the json file into the wiki. I did it once more. Still the `journal-date` field shows date in a format which is not same as in  Eric's solution and the date is of the created date, not the date of the Journal. I am using some other customisations in my wiki which uses the same date format, thats why I want that format of date

**Reply by buggyj** [reply]
change 

`time=<<now>>`

to
`
time=<<now "YYYY0MM0DD0hh0mm0ss000">>`

**Reply by arunnbabu81** [reply]
> 

I tried creating a Journal for August 20th. It shows journal-date or time field value of today's date, not August 20th . See the demo wiki

**Reply by buggyj** [reply]
That is what you ask for above...

I see in Eric's calendar (when creating a entry in the past) that journal-date = title-date + current-time  this is very illogical, why use the the current time???

**Reply by arunnbabu81** [reply]
> I see in Eric's calendar (when creating a entry in the past) that journal-date = title-date + current-time this is very illogical, why use the the current time???

I need the title-date of the every journal tiddlers because I have many other view template and customisation for journal tiddlers based on this date. Is there any way to achieve that.

**Reply by buggyj** [reply]
The problem is not with the date but with the time. 

I am trying to understand what time should be included when you create a journal with a date ''other than today''. 

If today at 13.00 I create a journal for jan 1 2000 then the date is "jan 1 2000" because I click on that date in the diary, however there are no times in the diary. - In Eric's plugin he seems to use the current time - this in my opinion is an error. (what relevance does the current time have with jan 1 2000?).

A time needs to be chosen (eg use default of zero hour zero minute zero second), or ignored completely.

**Reply by arunnbabu81** [reply]
> The problem is not with the date but with the time.
> 
> I am trying to understand what time should be included when you create a journal with a date ''other than today''.
> 
> If today at 13.00 I create a journal for jan 1 2000 then the date is "jan 1 2000" because I click on that date in the diary, however there are no times in the diary. - In Eric's plugin he seems to use the current time - this in my opinion is an error. (what relevance does the current time have with jan 1 2000?).
> 
> A time needs to be chosen (eg use default of zero hour zero minute zero second), or ignored completely.

Time actually doesnt matter for my purpose. Anytime will do.

**Reply by buggyj** [reply]
you can do this:

```
\define actions()
<$action-sendmessage $message="tm-new-tiddler" title=<<event-paramObject-title>> journal-date=<<event-paramObject-title>>>
\end
<$messagecatcher $tm-new-tiddler=<<actions>>>
<<diary>>
</$messagecatcher>
```

**Reply by arunnbabu81** [reply]
> you can do this:
> 
> ```
> \define actions()
> <$action-sendmessage $message="tm-new-tiddler" title=<<event-paramObject-title>> journal-date=<<event-paramObject-title>>>
> \end
> <$messagecatcher $tm-new-tiddler=<<actions>>>
> <<diary>>
> </$messagecatcher>
> ```

Is it possible to get the journal-date in this format - YYYY0MM0DD0hh0mm0ss000 or atleast YYYY0MM0DD

**Reply by buggyj** [reply]
I have yet to think of an easy general way to convert from a formatted string date back to a tiddlywiki 'date'. When a  journal is created it used the tiddler '$:/config/NewJournal/Title' to convert from the tiddlywiki internal date to a formatted string - there are many options, see https://tiddlywiki.com/#DateFormat. So the formatted date is used as the title of the journal - and this is, in general hard to convert to another format.

**Reply by buggyj** [reply]
you could change $:/config/NewJournal/Title to contain YYYY0MM0DD

**Reply by arunnbabu81** [reply]
> I have yet to think of an easy general way to convert from a formatted string date back to a tiddlywiki 'date'. When a journal is created it used the tiddler '$:/config/NewJournal/Title' to convert from the tiddlywiki internal date to a formatted string - there are many options, see https://tiddlywiki.com/#DateFormat. So the formatted date is used as the title of the journal - and this is, in general hard to convert to another format.

https://tiddlytools.com/timer.html#TiddlyTools%2FTime%2FParseDate

This is a macro used in Eric's solution to convert date. Is something similar possible.

**Reply by buggyj** [reply]
it is not possible in general to convert dates - for example 01/02/2022, this is the first of Feb in the English calendar but is the second of Jan in the American calendar

**Reply by arunnbabu81** [reply]
> it is not possible in general to convert dates - for example 01/02/2022, this is the first of Feb in the English calendar but is the second of Jan in the American calendar

Ok.

**Reply by buggyj** [reply]
Unfortunately I would have to write a new diary module to do this, and that would take me more time that I can spare at the moment as I am busy with other things.

**Reply by buggyj** [reply]
I have updated the calendar plugin,and a 'newdiary' macro. This is more flexible and can do what you are requesting.

https://bjcalendar.tiddlyhost.com/

**Reply by arunnbabu81** [reply]
> I have updated the calendar plugin,and a 'newdiary' macro. This is more flexible and can do what you are requesting.
> 
> https://bjcalendar.tiddlyhost.com/

Awesome @buggyj . Thank you for help me. I have two more small doubts
1. How to align this `calendarplus` calendar in the center of the tiddler.
2. Is there a way to toggle between year and month view in this `newdiary` macro similar to the sidebar calendar.

Once again thanks for taking your time to help me out.

**Reply by buggyj** [reply]
\define diarydets(i,tiddlerDate,rawDate,tags)
 <$ifnew tags="""$tags$""" dateraw=$rawDate$> [[$i$|$tiddlerDate$]]</$ifnew>
\end

@@display: flex;justify-content: center;
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="nomatch">
  <$button set="$:/temp/plugins/bj/Calendar" setTo="Year">Year</$button>
  <<newdiary "" "">>
</$reveal>
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="match">
  <$button set="$:/temp/plugins/bj/Calendar" setTo="Month">Month</$button>
<<newdiary 2014 "">>
</$reveal>
@@

**Reply by arunnbabu81** [reply]
It works. Is there a way to highlight the dates with already created Journal tidldler in the `Year` view, currently it works only for the `Month` view.

**Reply by buggyj** [reply]
It works for me - recheck that the dates are within the calendar  year.

**Reply by arunnbabu81** [reply]
> It works for me - recheck that the dates are within the calendar year.

It's working. I had forgot to change the year from 2014 to 2022. Sorry for the confusion.

**Reply by buggyj** [reply]
This is better:

```
\define diarydets(i,tiddlerDate,rawDate,tags)
<$ifnew tags="""$tags$""" dateraw=$rawDate$> [[$i$|$tiddlerDate$]]</$ifnew>
\end

@@display: flex;justify-content: center;
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="nomatch">
<$button set="$:/temp/plugins/bj/Calendar" setTo="Year">Year</$button>
<<newdiary>>
</$reveal>
<$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="match">
<$button set="$:/temp/plugins/bj/Calendar" setTo="Month">Month</$button>
<$macrocall $name=newdiary year =<<now YYYY>>/>
</$reveal>
@@
```

**Reply by arunnbabu81** [reply]
> This is better:
> 
> ```
> \define diarydets(i,tiddlerDate,rawDate,tags)
> <$ifnew tags="""$tags$""" dateraw=$rawDate$> [[$i$|$tiddlerDate$]]</$ifnew>
> \end
> 
> @@display: flex;justify-content: center;
> <$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="nomatch">
> <$button set="$:/temp/plugins/bj/Calendar" setTo="Year">Year</$button>
> <<newdiary>>
> </$reveal>
> <$reveal state="$:/temp/plugins/bj/Calendar" text="Year" type="match">
> <$button set="$:/temp/plugins/bj/Calendar" setTo="Month">Month</$button>
> <$macrocall $name=newdiary year =<<now YYYY>>/>
> </$reveal>
> @@
> ```

Thank you.
